hawkbit-repository/hawkbit-repository-jpa-eclipselink/src/main/java/org/eclipse/hawkbit/repository/jpa/MultiTenantJpaTransactionManager.java

```diff
@@ -0,0 +1,91 @@
+
+import org.eclipse.hawkbit.repository.jpa.model.EntityPropertyChangeListener;
+import org.eclipse.hawkbit.tenancy.TenantAware;
+import org.eclipse.persistence.config.PersistenceUnitProperties;
+import org.eclipse.persistence.descriptors.ClassDescriptor;
+import org.eclipse.persistence.descriptors.DescriptorEventManager;
+import org.eclipse.persistence.sessions.Session;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.orm.jpa.EntityManagerHolder;
+import org.springframework.orm.jpa.JpaTransactionManager;
+import org.springframework.transaction.TransactionDefinition;
+import org.springframework.transaction.support.TransactionSynchronizationManager;
+
+/**
+ * {@link JpaTransactionManager} that sets the {@link TenantAware#getCurrentTenant()} in the eclipselink session. This has
+ * to be done in eclipselink after a {@link Transaction} has been started.
+ * <p/>
+ * The class also handles setting the {@link EntityPropertyChangeListener} to the {@link DescriptorEventManager} of the
+ */
+class MultiTenantJpaTransactionManager extends JpaTransactionManager {
+
+    @Serial
+    private static final long serialVersionUID = 1L;
+
+    @Autowired
+    private transient TenantAware tenantAware;
+
+    private static final Class<?> JPA_TARGET;
+
+    static {
+        try {
+            JPA_TARGET = Class.forName("org.eclipse.hawkbit.repository.jpa.model.JpaTarget");
+        } catch (final RuntimeException e) {
+            throw e;
+        } catch (final Exception e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    private static final EntityPropertyChangeListener ENTITY_PROPERTY_CHANGE_LISTENER = new EntityPropertyChangeListener();
+
+    @Override
+    protected void doBegin(final Object transaction, final TransactionDefinition definition) {
+        super.doBegin(transaction, definition);
+
+        final EntityManager em = Objects.requireNonNull(
+                        (EntityManagerHolder) TransactionSynchronizationManager.getResource(
+                                Objects.requireNonNull(
+                                        getEntityManagerFactory(),
+                                        "No EntityManagerFactory provided by TransactionSynchronizationManager")),
+                        "No EntityManagerHolder provided by TransactionSynchronizationManager")
+                .getEntityManager();
+
+        final ClassDescriptor classDescriptor = em.unwrap(Session.class).getClassDescriptor(JPA_TARGET);
+        if (classDescriptor != null) {
+            final DescriptorEventManager dem = classDescriptor.getEventManager();
+            if (dem != null && !dem.getEventListeners().contains(ENTITY_PROPERTY_CHANGE_LISTENER)) {
+                dem.addListener(ENTITY_PROPERTY_CHANGE_LISTENER);
+            }
+        }
+
+        final String currentTenant = tenantAware.getCurrentTenant();
+        if (currentTenant == null) {
+            cleanupTenant(em);
```

Comment:
don't see this cleanup logic when currentTenant is null in previous eclipse link impl?

Reply:
it's just for sure .. in some cases seems that entity manager is release and hence tenant .. however if reused for some reasons - let clean it in that case